In [1]:
from pathlib import Path

project = Path.cwd().parent
raw_data = project / "raw data"

print("Project:", project)
print("Raw data exists:", raw_data.exists())

print("\nFiles in raw data:")
for file in raw_data.iterdir():
    print(file.name)

Project: /Users/dikshadev/    Projects/NYC_HV_FHV_ Analytics_2026
Raw data exists: True

Files in raw data:
fhvhv_tripdata_2026-01.parquet
.DS_Store
fhvhv_tripdata_2026-02.parquet
fhvhv_tripdata_2026-03.parquet


In [2]:
import pandas as pd

jan_file = raw_data / "fhvhv_tripdata_2026-01.parquet"

df = pd.read_parquet(jan_file)

print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 20940373
Columns: 25


In [3]:
df.columns.tolist()

['hvfhs_license_num',
 'dispatching_base_num',
 'originating_base_num',
 'request_datetime',
 'on_scene_datetime',
 'pickup_datetime',
 'dropoff_datetime',
 'PULocationID',
 'DOLocationID',
 'trip_miles',
 'trip_time',
 'base_passenger_fare',
 'tolls',
 'bcf',
 'sales_tax',
 'congestion_surcharge',
 'airport_fee',
 'tips',
 'driver_pay',
 'shared_request_flag',
 'shared_match_flag',
 'access_a_ride_flag',
 'wav_request_flag',
 'wav_match_flag',
 'cbd_congestion_fee']

In [4]:
df.dtypes

hvfhs_license_num                  str
dispatching_base_num               str
originating_base_num               str
request_datetime        datetime64[us]
on_scene_datetime       datetime64[us]
pickup_datetime         datetime64[us]
dropoff_datetime        datetime64[us]
PULocationID                     int32
DOLocationID                     int32
trip_miles                     float64
trip_time                        int64
base_passenger_fare            float64
tolls                          float64
bcf                            float64
sales_tax                      float64
congestion_surcharge           float64
airport_fee                    float64
tips                           float64
driver_pay                     float64
shared_request_flag                str
shared_match_flag                  str
access_a_ride_flag                 str
wav_request_flag                   str
wav_match_flag                     str
cbd_congestion_fee             float64
dtype: object

In [5]:
df.isnull().sum()

hvfhs_license_num             0
dispatching_base_num          0
originating_base_num    5676843
request_datetime              0
on_scene_datetime             0
pickup_datetime               0
dropoff_datetime              0
PULocationID                  0
DOLocationID                  0
trip_miles                    0
trip_time                     0
base_passenger_fare           0
tolls                         0
bcf                           0
sales_tax                     0
congestion_surcharge          0
airport_fee                   0
tips                          0
driver_pay                    0
shared_request_flag           0
shared_match_flag             0
access_a_ride_flag            0
wav_request_flag              0
wav_match_flag                0
cbd_congestion_fee            0
dtype: int64

In [6]:
df.duplicated().sum()

np.int64(0)

In [7]:
df.describe().T

,count,mean,min,25%,50%,75%,max,std
request_datetime,20940373,2026-01-17 01:16:50.399517,2025-12-31 23:00:00,2026-01-09 15:54:46,2026-01-17 03:06:17,2026-01-24 12:35:52,2026-02-01 00:43:00,NaN
on_scene_datetime,20940373,2026-01-17 01:20:54.320515,2025-12-31 23:07:56,2026-01-09 15:57:57,2026-01-17 03:10:22,2026-01-24 12:39:07,2026-01-31 23:59:56,NaN
pickup_datetime,20940373,2026-01-17 01:21:53.479783,2026-01-01 00:00:00,2026-01-09 15:58:58,2026-01-17 03:11:20,2026-01-24 12:40:03,2026-01-31 23:59:59,NaN
dropoff_datetime,20940373,2026-01-17 01:40:38.742369,2026-01-01 00:00:56,2026-01-09 16:20:25,2026-01-17 03:28:16,2026-01-24 12:57:07,2026-02-01 04:46:31,NaN
PULocationID,20940373.0,137.643308,1.0,74.0,138.0,209.0,265.0,75.335746
DOLocationID,20940373.0,141.175015,1.0,75.0,140.0,216.0,265.0,78.387038
trip_miles,20940373.0,4.637687,0.0,1.41,2.69,5.73,1454.43,5.556571
trip_time,20940373.0,1125.222466,0.0,570.0,921.0,1453.0,44998.0,792.208868
base_passenger_fare,20940373.0,25.651055,-260.9,11.64,18.53,30.58,1596.1,24.02356
tolls,20940373.0,0.982117,0.0,0.0,0.0,0.0,108.41,3.427288


In [8]:
print("Trip miles > 100:", (df["trip_miles"] > 100).sum())
print("Trip time > 4 hours:", (df["trip_time"] > 14400).sum())
print("Zero-mile trips:", (df["trip_miles"] == 0).sum())
print("Zero-time trips:", (df["trip_time"] == 0).sum())
print("Negative passenger fares:", (df["base_passenger_fare"] < 0).sum())
print("Negative driver pay:", (df["driver_pay"] < 0).sum())

Trip miles > 100: 1627
Trip time > 4 hours: 255
Zero-mile trips: 2782
Zero-time trips: 737
Negative passenger fares: 29965
Negative driver pay: 56


In [9]:
df.loc[
    df["base_passenger_fare"] < 0,
    [
        "request_datetime",
        "pickup_datetime",
        "dropoff_datetime",
        "trip_miles",
        "trip_time",
        "base_passenger_fare",
        "tolls",
        "sales_tax",
        "congestion_surcharge",
        "airport_fee",
        "tips",
        "driver_pay"
    ]
].head(20)

,request_datetime,pickup_datetime,dropoff_datetime,trip_miles,trip_time,base_passenger_fare,tolls,sales_tax,congestion_surcharge,airport_fee,tips,driver_pay
16178,2026-01-01 00:41:45,2026-01-01 00:52:50,2026-01-01 01:03:07,2.84,617,-1.63,0.00,1.28,0.00,0.0,0.00,12.52
18858,2026-01-01 00:23:36,2026-01-01 00:37:48,2026-01-01 00:44:25,0.24,397,-3.50,0.00,2.73,0.00,0.0,0.00,20.86
24107,2026-01-01 00:38:31,2026-01-01 00:49:10,2026-01-01 01:51:41,15.63,3751,-70.46,0.00,34.65,2.75,0.0,0.00,196.13
33496,2026-01-01 00:06:17,2026-01-01 00:42:26,2026-01-01 00:46:54,0.06,268,-0.79,0.00,0.62,0.00,0.0,0.00,16.00
64653,2026-01-01 00:55:48,2026-01-01 01:07:09,2026-01-01 01:14:12,0.11,423,-3.47,0.00,2.71,0.00,0.0,0.00,23.53
69352,2026-01-01 01:29:19,2026-01-01 01:33:31,2026-01-01 02:17:29,11.94,2639,-18.27,16.06,0.00,0.00,0.0,10.72,59.78
70403,2026-01-01 01:17:24,2026-01-01 01:23:57,2026-01-01 01:29:38,0.18,340,-3.13,0.00,2.44,0.00,0.0,0.00,16.90
71017,2026-01-01 01:14:44,2026-01-01 01:20:41,2026-01-01 01:50:37,15.56,1796,-23.38,16.06,0.00,0.00,0.0,0.00,147.71
76296,2026-01-01 01:12:41,2026-01-01 01:15:26,2026-01-01 01:20:31,0.97,306,-1.39,0.00,1.09,0.00,0.0,0.00,7.58
80799,2026-01-01 01:25:18,2026-01-01 01:30:28,2026-01-01 01:43:29,2.79,781,-4.65,0.00,1.36,0.00,0.0,2.00,14.87


In [10]:
df.loc[df["base_passenger_fare"] < 0, "base_passenger_fare"].describe()

count    29965.000000
mean        -1.842493
std          3.233205
min       -260.900000
25%         -3.220000
50%         -0.770000
75%         -0.420000
max         -0.010000
Name: base_passenger_fare, dtype: float64

In [11]:
df.loc[df["base_passenger_fare"] < 0].groupby(
    "hvfhs_license_num"
).size()

hvfhs_license_num
HV0003    29934
HV0005       31
dtype: int64

In [12]:
df.loc[
    df["base_passenger_fare"] < 0,
    [
        "trip_miles",
        "trip_time",
        "base_passenger_fare",
        "driver_pay",
        "tips"
    ]
].describe()

,trip_miles,trip_time,base_passenger_fare,driver_pay,tips
count,29965.000000,29965.000000,29965.000000,29965.000000,29965.000000
mean,3.457255,1075.033472,-1.842493,2.343767,0.235239
std,4.743829,1094.572978,3.233205,11.059135,1.453081
min,0.000000,1.000000,-260.900000,-14.890000,0.000000
25%,1.140000,510.000000,-3.220000,0.000000,0.000000
50%,1.970000,808.000000,-0.770000,0.000000,0.000000
75%,4.010000,1337.000000,-0.420000,0.000000,0.000000
max,273.960000,39734.000000,-0.010000,901.460000,50.780000


In [13]:
df.loc[
    df["base_passenger_fare"] < 0,
    [
        "shared_request_flag",
        "shared_match_flag",
        "access_a_ride_flag",
        "wav_request_flag",
        "wav_match_flag"
    ]
].apply(pd.Series.value_counts)

,shared_request_flag,shared_match_flag,access_a_ride_flag,wav_request_flag,wav_match_flag
N,28427,29291,29959,29934,26569
Y,1538,674,6,31,3396


In [14]:
negative_fares = df[df["base_passenger_fare"] < 0]

print("Negative fare records:", len(negative_fares))
print("Driver pay = $0:", (negative_fares["driver_pay"] == 0).sum())
print("Driver pay > $0:", (negative_fares["driver_pay"] > 0).sum())
print("Driver pay < $0:", (negative_fares["driver_pay"] < 0).sum())

Negative fare records: 29965
Driver pay = $0: 24279
Driver pay > $0: 5671
Driver pay < $0: 15


In [15]:
print("Zero-mile trips:", (df["trip_miles"] == 0).sum())
print("Zero-mile + zero-time trips:",
      ((df["trip_miles"] == 0) & (df["trip_time"] == 0)).sum())
print("Zero-mile + positive-time trips:",
      ((df["trip_miles"] == 0) & (df["trip_time"] > 0)).sum())

Zero-mile trips: 2782
Zero-mile + zero-time trips: 737
Zero-mile + positive-time trips: 2045


In [16]:
df.loc[
    df["trip_miles"] == 0,
    [
        "trip_time",
        "base_passenger_fare",
        "driver_pay",
        "tips"
    ]
].describe()

,trip_time,base_passenger_fare,driver_pay,tips
count,2782.000000,2782.000000,2782.000000,2782.000000
mean,253.189073,14.539543,10.759457,0.465658
std,650.504759,13.882911,10.579042,2.377058
min,0.000000,-45.460000,0.000000,0.000000
25%,0.000000,7.220000,4.280000,0.000000
50%,149.000000,10.000000,6.975000,0.000000
75%,298.000000,17.182500,12.617500,0.000000
max,21176.000000,232.440000,153.140000,50.000000


In [17]:
zero_both = df[
    (df["trip_miles"] == 0) &
    (df["trip_time"] == 0)
]

zero_both[
    [
        "base_passenger_fare",
        "driver_pay",
        "tips",
        "shared_request_flag",
        "shared_match_flag"
    ]
].describe()

,base_passenger_fare,driver_pay,tips
count,737.000000,737.000000,737.000000
mean,24.767436,18.600326,0.788494
std,17.717712,13.676089,2.678562
min,0.000000,0.000000,0.000000
25%,12.520000,8.310000,0.000000
50%,19.730000,14.680000,0.000000
75%,32.300000,25.560000,0.000000
max,123.710000,104.850000,27.360000


In [18]:
df.loc[
    df["trip_miles"] > 100,
    [
        "trip_miles",
        "trip_time",
        "base_passenger_fare",
        "driver_pay"
    ]
].describe()

,trip_miles,trip_time,base_passenger_fare,driver_pay
count,1627.000000,1627.000000,1627.000000,1627.000000
mean,131.492461,9811.748617,436.955095,337.409766
std,68.247496,3559.516792,176.274733,124.541205
min,100.010000,1188.000000,-260.900000,0.000000
25%,106.875000,7778.000000,333.050000,264.410000
50%,115.430000,8886.000000,390.360000,300.220000
75%,133.035000,10629.000000,485.575000,361.780000
max,1454.430000,44998.000000,1596.100000,1243.910000


In [19]:
df.loc[
    df["trip_miles"] > 100,
    [
        "hvfhs_license_num",
        "pickup_datetime",
        "dropoff_datetime",
        "PULocationID",
        "DOLocationID",
        "trip_miles",
        "trip_time",
        "base_passenger_fare",
        "driver_pay"
    ]
].sort_values("trip_miles", ascending=False).head(20)

,hvfhs_license_num,pickup_datetime,dropoff_datetime,PULocationID,DOLocationID,trip_miles,trip_time,base_passenger_fare,driver_pay
7036830,HV0003,2026-01-11 22:50:05,2026-01-11 23:40:12,138,265,1454.430,3007,181.76,127.35
7023842,HV0003,2026-01-11 22:59:40,2026-01-11 23:19:28,138,164,1424.980,1188,64.03,58.32
7014472,HV0003,2026-01-11 22:49:49,2026-01-11 23:20:06,138,142,1423.970,1817,69.43,54.84
13037649,HV0003,2026-01-20 19:33:23,2026-01-21 05:48:43,197,197,571.310,36920,1288.51,1115.80
8155354,HV0003,2026-01-13 19:42:31,2026-01-14 05:27:56,197,197,565.740,35125,1284.55,1088.81
9350955,HV0003,2026-01-15 17:15:39,2026-01-16 03:33:09,197,197,540.030,37050,1313.92,1243.91
18722003,HV0003,2026-01-29 12:23:15,2026-01-29 19:42:31,252,53,343.600,26355,916.69,715.94
670864,HV0003,2026-01-01 21:20:09,2026-01-02 03:58:18,188,188,340.480,23889,738.28,685.80
19799039,HV0005,2026-01-30 19:50:31,2026-01-31 02:45:43,61,265,327.038,24912,1596.10,840.85
19629020,HV0005,2026-01-30 15:03:32,2026-01-30 18:27:50,137,265,283.613,12258,663.45,626.34


In [20]:
df["avg_speed_mph"] = df["trip_miles"] / (df["trip_time"] / 3600)

df["avg_speed_mph"].describe()

count    2.093964e+07
mean     1.311361e+01
std      7.212190e+00
min      0.000000e+00
25%      8.307692e+00
50%      1.100938e+01
75%      1.592195e+01
max      4.318121e+03
Name: avg_speed_mph, dtype: float64

In [21]:
print("Speed > 100 mph:", (df["avg_speed_mph"] > 100).sum())
print("Speed > 150 mph:", (df["avg_speed_mph"] > 150).sum())
print("Speed > 200 mph:", (df["avg_speed_mph"] > 200).sum())
print("Speed > 500 mph:", (df["avg_speed_mph"] > 500).sum())

Speed > 100 mph: 13
Speed > 150 mph: 7
Speed > 200 mph: 6
Speed > 500 mph: 4


In [22]:
df.loc[
    df["avg_speed_mph"] > 100,
    [
        "hvfhs_license_num",
        "pickup_datetime",
        "dropoff_datetime",
        "PULocationID",
        "DOLocationID",
        "trip_miles",
        "trip_time",
        "avg_speed_mph",
        "base_passenger_fare",
        "driver_pay"
    ]
].sort_values("avg_speed_mph", ascending=False)

,hvfhs_license_num,pickup_datetime,dropoff_datetime,PULocationID,DOLocationID,trip_miles,trip_time,avg_speed_mph,base_passenger_fare,driver_pay
7023842,HV0003,2026-01-11 22:59:40,2026-01-11 23:19:28,138,164,1424.980,1188,4318.121212,64.03,58.32
7014472,HV0003,2026-01-11 22:49:49,2026-01-11 23:20:06,138,142,1423.970,1817,2821.294441,69.43,54.84
7036830,HV0003,2026-01-11 22:50:05,2026-01-11 23:40:12,138,265,1454.430,3007,1741.253076,181.76,127.35
17654946,HV0003,2026-01-27 22:22:26,2026-01-27 22:25:33,265,265,34.590,186,669.483871,267.34,133.65
10833207,HV0003,2026-01-17 16:46:28,2026-01-17 18:03:56,109,132,21.290,226,339.132743,119.08,54.82
8633618,HV0003,2026-01-14 16:08:15,2026-01-14 17:39:28,142,132,13.770,195,254.215385,159.68,103.39
1410740,HV0005,2026-01-03 04:58:17,2026-01-03 04:58:18,219,219,0.043,1,154.800000,6.54,4.00
3153625,HV0005,2026-01-06 06:41:05,2026-01-06 06:41:06,117,117,0.035,1,126.000000,7.30,4.00
10943426,HV0003,2026-01-17 18:20:01,2026-01-17 19:23:36,138,265,130.410,3815,123.060550,105.20,102.09
6960317,HV0005,2026-01-11 20:28:55,2026-01-11 20:28:56,112,80,0.034,1,122.400000,10.07,4.00


In [23]:
print(
    "Request after on-scene:",
    (df["request_datetime"] > df["on_scene_datetime"]).sum()
)

print(
    "On-scene after pickup:",
    (df["on_scene_datetime"] > df["pickup_datetime"]).sum()
)

print(
    "Pickup after dropoff:",
    (df["pickup_datetime"] > df["dropoff_datetime"]).sum()
)

Request after on-scene: 356754
On-scene after pickup: 694
Pickup after dropoff: 0


In [24]:
id="qv3r8a"
df.loc[
    df["request_datetime"] > df["on_scene_datetime"],
    [
        "request_datetime",
        "on_scene_datetime",
        "pickup_datetime",
        "dropoff_datetime",
        "trip_miles",
        "trip_time"
    ]
].head(20)

,request_datetime,on_scene_datetime,pickup_datetime,dropoff_datetime,trip_miles,trip_time
87,2026-01-01 01:00:01,2026-01-01 00:56:18,2026-01-01 00:57:40,2026-01-01 01:30:57,6.954,1997
245,2026-01-01 00:45:00,2026-01-01 00:42:00,2026-01-01 00:48:10,2026-01-01 01:27:23,4.070,2354
301,2026-01-01 00:30:01,2026-01-01 00:22:28,2026-01-01 00:30:00,2026-01-01 00:55:24,6.145,1524
378,2026-01-01 00:25:01,2026-01-01 00:17:44,2026-01-01 00:26:00,2026-01-01 00:52:12,4.740,1572
597,2026-01-01 00:40:00,2026-01-01 00:26:53,2026-01-01 00:28:39,2026-01-01 00:40:00,3.290,680
617,2026-01-01 01:00:01,2026-01-01 00:53:35,2026-01-01 00:56:49,2026-01-01 01:36:06,9.368,2357
758,2026-01-01 00:30:00,2026-01-01 00:27:01,2026-01-01 00:30:22,2026-01-01 01:14:13,7.030,2630
801,2026-01-01 00:30:01,2026-01-01 00:22:21,2026-01-01 00:22:39,2026-01-01 00:32:54,1.929,615
805,2026-01-01 00:15:01,2026-01-01 00:06:14,2026-01-01 00:15:00,2026-01-01 00:35:41,3.361,1241
823,2026-01-01 00:30:01,2026-01-01 00:24:53,2026-01-01 00:30:00,2026-01-01 00:45:24,2.675,924


In [25]:
request_on_scene_diff = (
    df["request_datetime"] - df["on_scene_datetime"]
)

print("Median difference:", request_on_scene_diff.median())
print("Mean difference:", request_on_scene_diff.mean())
print("Minimum difference:", request_on_scene_diff.min())
print("Maximum difference:", request_on_scene_diff.max())

Median difference: -1 days +23:56:46
Mean difference: -1 days +23:55:56.079008
Minimum difference: -2 days +21:01:10
Maximum difference: 0 days 09:10:13


In [26]:
positive_diff = request_on_scene_diff[
    request_on_scene_diff > pd.Timedelta(0)
]

print("Positive differences:", len(positive_diff))

print("\nPercentiles of positive differences:")
print(positive_diff.quantile([0.25, 0.50, 0.75, 0.90, 0.95, 0.99]))

Positive differences: 356754

Percentiles of positive differences:
0.25   0 days 00:05:21
0.50   0 days 00:08:35
0.75   0 days 00:13:04
0.90   0 days 00:18:31
0.95   0 days 00:21:05
0.99   0 days 00:33:38
dtype: timedelta64[us]


In [27]:
df.loc[
    df["trip_time"] > 14400,
    [
        "hvfhs_license_num",
        "pickup_datetime",
        "dropoff_datetime",
        "trip_miles",
        "trip_time",
        "avg_speed_mph",
        "base_passenger_fare",
        "driver_pay"
    ]
].sort_values("trip_time", ascending=False).head(20)

,hvfhs_license_num,pickup_datetime,dropoff_datetime,trip_miles,trip_time,avg_speed_mph,base_passenger_fare,driver_pay
16397978,HV0003,2026-01-25 07:44:47,2026-01-25 20:14:44,228.150,44998,18.252811,0.00,0.00
16122914,HV0003,2026-01-24 21:15:03,2026-01-25 08:17:17,21.240,39734,1.924397,-79.52,129.04
9350955,HV0003,2026-01-15 17:15:39,2026-01-16 03:33:09,540.030,37050,52.472551,1313.92,1243.91
13037649,HV0003,2026-01-20 19:33:23,2026-01-21 05:48:43,571.310,36920,55.707367,1288.51,1115.80
11591979,HV0005,2026-01-18 15:00:17,2026-01-19 00:57:58,156.042,35861,15.664683,1333.33,1016.17
2398066,HV0005,2026-01-04 18:16:33,2026-01-05 04:11:48,111.356,35715,11.224460,1171.49,878.60
8155354,HV0003,2026-01-13 19:42:31,2026-01-14 05:27:56,565.740,35125,57.983317,1284.55,1088.81
3139151,HV0005,2026-01-06 05:16:21,2026-01-06 15:00:23,80.832,35042,8.304184,1034.80,776.10
17705326,HV0003,2026-01-28 00:58:46,2026-01-28 10:40:18,36.510,34892,3.766938,-53.11,132.21
19714920,HV0005,2026-01-30 17:22:37,2026-01-31 03:01:52,84.792,34755,8.782943,1045.11,783.83


In [28]:
timestamp_trip_time = (
    df["dropoff_datetime"] - df["pickup_datetime"]
).dt.total_seconds()

time_difference = timestamp_trip_time - df["trip_time"]

print("Median difference:", time_difference.median())
print("Mean difference:", time_difference.mean())
print("Minimum difference:", time_difference.min())
print("Maximum difference:", time_difference.max())

Median difference: 0.0
Mean difference: 0.040121439909403714
Minimum difference: -3.0
Maximum difference: 10755.0


In [29]:
print("Difference > 1 minute:", (time_difference.abs() > 60).sum())
print("Difference > 5 minutes:", (time_difference.abs() > 300).sum())
print("Difference > 10 minutes:", (time_difference.abs() > 600).sum())

Difference > 1 minute: 1333
Difference > 5 minutes: 1250
Difference > 10 minutes: 922


In [30]:
df.loc[
    df["originating_base_num"].isna()
].groupby("hvfhs_license_num").size()

hvfhs_license_num
HV0005    5676843
dtype: int64

In [31]:
df.groupby("hvfhs_license_num").size()

hvfhs_license_num
HV0003    15245631
HV0005     5694742
dtype: int64

In [32]:
financial_columns = [
    "tolls",
    "bcf",
    "sales_tax",
    "congestion_surcharge",
    "airport_fee",
    "tips",
    "cbd_congestion_fee"
]

for col in financial_columns:
    print(f"{col}: {(df[col] < 0).sum()} negative values")

tolls: 0 negative values
bcf: 0 negative values
sales_tax: 0 negative values
congestion_surcharge: 0 negative values
airport_fee: 0 negative values
tips: 0 negative values
cbd_congestion_fee: 0 negative values


## Data Quality Findings & Cleaning Decisions

### Duplicate Records
- No exact duplicate rows were found.
- **Action:** Retain all records.

### Missing Values
- `originating_base_num` has 5,676,843 missing values.
- 99.7% of HV0005 records have this field missing, while HV0003 has complete coverage.
- **Action:** Retain missing values; no imputation was performed because the missingness is systematic.

### Negative Financial Values
- `base_passenger_fare`: 29,965 negative values.
- `driver_pay`: 56 negative values.
- Other financial fields contain no negative values.
- **Action:** Retain these records and create flags rather than removing them. They may represent financial adjustments or other source-system records and should not be assumed invalid without further documentation.

### Zero-Distance Trips
- 2,782 trips have `trip_miles = 0`.
- 737 trips have both `trip_miles = 0` and `trip_time = 0`.
- These records can contain positive fares and driver pay.
- **Action:** Retain them.

### Extreme Speed / Distance Anomalies
- 13 trips have an implied average speed above 100 mph.
- These records contain internally implausible distance-duration combinations.
- **Action:** Retain the original records but flag them as speed/distance anomalies. They can be excluded from analyses where distance or speed is relevant.

### Timestamp Validation
- No records have `pickup_datetime > dropoff_datetime`.
- `trip_time` is highly consistent with the difference between pickup and dropoff timestamps.
- **Action:** Retain the timestamp and trip-duration fields.

### Long Trips
- 255 trips have `trip_time > 4 hours`.
- Several are plausible long-distance trips.
- **Action:** Do not remove based solely on trip duration.

### Raw Data
- The original January Parquet file remains unchanged.
- Cleaning and derived fields will be created in the `processed data` folder.

In [33]:
import numpy as np

# Create a working copy of the raw data
processed_df = df.copy()

# --------------------------------------------------
# 1. Trip duration in minutes
# --------------------------------------------------
processed_df["trip_duration_minutes"] = (
    processed_df["trip_time"] / 60
)

# --------------------------------------------------
# 2. Average speed in mph
# --------------------------------------------------
processed_df["avg_speed_mph"] = np.where(
    processed_df["trip_time"] > 0,
    processed_df["trip_miles"] / (processed_df["trip_time"] / 3600),
    np.nan
)

# --------------------------------------------------
# 3. Data quality flags
# --------------------------------------------------

# Negative financial values
processed_df["negative_fare_flag"] = (
    processed_df["base_passenger_fare"] < 0
)

processed_df["negative_driver_pay_flag"] = (
    processed_df["driver_pay"] < 0
)

# Zero-distance trips
processed_df["zero_distance_flag"] = (
    processed_df["trip_miles"] == 0
)

# Zero-distance AND zero-duration trips
processed_df["zero_distance_zero_time_flag"] = (
    (processed_df["trip_miles"] == 0) &
    (processed_df["trip_time"] == 0)
)

# Very high implied speed
processed_df["high_speed_flag"] = (
    processed_df["avg_speed_mph"] > 100
)

# Very long trips
processed_df["long_trip_flag"] = (
    processed_df["trip_time"] > 14400
)

# Missing originating base
processed_df["missing_originating_base_flag"] = (
    processed_df["originating_base_num"].isna()
)

# --------------------------------------------------
# 4. Timestamp consistency flag
# --------------------------------------------------
processed_df["invalid_timestamp_flag"] = (
    processed_df["pickup_datetime"] > processed_df["dropoff_datetime"]
)

print("Processed DataFrame created successfully.")
print("Rows:", processed_df.shape[0])
print("Columns:", processed_df.shape[1])

Processed DataFrame created successfully.
Rows: 20940373
Columns: 35


In [34]:
flag_columns = [
    "negative_fare_flag",
    "negative_driver_pay_flag",
    "zero_distance_flag",
    "zero_distance_zero_time_flag",
    "high_speed_flag",
    "long_trip_flag",
    "missing_originating_base_flag",
    "invalid_timestamp_flag"
]

for col in flag_columns:
    print(f"{col}: {processed_df[col].sum()}")

negative_fare_flag: 29965
negative_driver_pay_flag: 56
zero_distance_flag: 2782
zero_distance_zero_time_flag: 737
high_speed_flag: 13
long_trip_flag: 255
missing_originating_base_flag: 5676843
invalid_timestamp_flag: 0


In [35]:
# Trip demand by pickup hour

hourly_demand = (
    processed_df
    .assign(pickup_hour=processed_df["pickup_datetime"].dt.hour)
    .groupby("pickup_hour")
    .size()
    .reset_index(name="trip_count")
)

hourly_demand

,pickup_hour,trip_count
0,0,729999
1,1,515768
2,2,386609
3,3,313318
4,4,321638
5,5,377950
6,6,582924
7,7,886284
8,8,1087379
9,9,990742


In [36]:
# Identify peak and lowest demand hours

peak_hour = hourly_demand.loc[
    hourly_demand["trip_count"].idxmax()
]

lowest_hour = hourly_demand.loc[
    hourly_demand["trip_count"].idxmin()
]

print("Peak demand hour:")
print(peak_hour)

print("\nLowest demand hour:")
print(lowest_hour)

Peak demand hour:
pickup_hour         18
trip_count     1275537
Name: 18, dtype: int64

Lowest demand hour:
pickup_hour         3
trip_count     313318
Name: 3, dtype: int64


In [37]:
peak_count = peak_hour["trip_count"]
lowest_count = lowest_hour["trip_count"]

difference = peak_count - lowest_count
multiple = peak_count / lowest_count

print(f"Difference in trips: {difference:,}")
print(f"Peak hour is {multiple:.2f}x the lowest-demand hour.")

Difference in trips: 962,219
Peak hour is 4.07x the lowest-demand hour.


In [38]:
# Trip demand by day of week

daily_demand = (
    processed_df
    .assign(
        day_of_week=processed_df["pickup_datetime"].dt.day_name()
    )
    .groupby("day_of_week")
    .size()
    .reindex([
        "Monday",
        "Tuesday",
        "Wednesday",
        "Thursday",
        "Friday",
        "Saturday",
        "Sunday"
    ])
    .reset_index(name="trip_count")
)

daily_demand

,day_of_week,trip_count
0,Monday,2199133
1,Tuesday,2536803
2,Wednesday,2606897
3,Thursday,3508966
4,Friday,3754036
5,Saturday,4045717
6,Sunday,2288821


In [39]:
# Trip demand by provider

provider_demand = (
    processed_df
    .groupby("hvfhs_license_num")
    .size()
    .reset_index(name="trip_count")
    .sort_values("trip_count", ascending=False)
)

provider_demand

,hvfhs_license_num,trip_count
0,HV0003,15245631
1,HV0005,5694742


In [40]:
# Trip demand by pickup location

pickup_demand = (
    processed_df
    .groupby("PULocationID")
    .size()
    .reset_index(name="trip_count")
    .sort_values("trip_count", ascending=False)
)

pickup_demand.head(20)

,PULocationID,trip_count
135,138,364872
129,132,330578
60,61,277903
36,37,246773
75,76,245850
78,79,242997
227,230,231846
158,161,226497
178,181,214217
228,231,213850


In [41]:
# Trip demand by drop-off location

dropoff_demand = (
    processed_df
    .groupby("DOLocationID")
    .size()
    .reset_index(name="trip_count")
    .sort_values("trip_count", ascending=False)
)

dropoff_demand.head(20)

,DOLocationID,trip_count
261,265,847744
135,138,363249
129,132,357339
60,61,294175
36,37,249946
75,76,244871
178,181,216007
109,112,207670
78,79,207358
6,7,204689


In [42]:
# Trip distance distribution

processed_df["trip_miles"].describe()

count    2.094037e+07
mean     4.637687e+00
std      5.556571e+00
min      0.000000e+00
25%      1.410000e+00
50%      2.690000e+00
75%      5.730000e+00
max      1.454430e+03
Name: trip_miles, dtype: float64

In [43]:
# Trip duration distribution

processed_df["trip_time"].describe()

count    2.094037e+07
mean     1.125222e+03
std      7.922089e+02
min      0.000000e+00
25%      5.700000e+02
50%      9.210000e+02
75%      1.453000e+03
max      4.499800e+04
Name: trip_time, dtype: float64

In [44]:
# Average trip speed by pickup hour

hourly_speed = (
    processed_df
    .assign(pickup_hour=processed_df["pickup_datetime"].dt.hour)
    .groupby("pickup_hour")["avg_speed_mph"]
    .mean()
    .reset_index()
    .sort_values("pickup_hour")
)

hourly_speed

,pickup_hour,avg_speed_mph
0,0,15.356880
1,1,15.507842
2,2,16.051297
3,3,17.617471
4,4,19.370292
5,5,19.044807
6,6,16.396221
7,7,13.311710
8,8,12.075672
9,9,12.592692


In [45]:
# Average trip duration by pickup hour

hourly_duration = (
    processed_df
    .assign(pickup_hour=processed_df["pickup_datetime"].dt.hour)
    .groupby("pickup_hour")["trip_time"]
    .mean()
    .reset_index()
    .sort_values("pickup_hour")
)

hourly_duration

,pickup_hour,trip_time
0,0,1010.528776
1,1,985.073828
2,2,983.360067
3,3,1011.966526
4,4,1046.841303
5,5,1107.419473
6,6,1143.105919
7,7,1159.500794
8,8,1133.556525
9,9,1107.389356


In [46]:
# Fare and driver pay distribution

processed_df[
    ["base_passenger_fare", "driver_pay", "tips"]
].describe()

,base_passenger_fare,driver_pay,tips
count,2.094037e+07,2.094037e+07,2.094037e+07
mean,2.565105e+01,1.983287e+01,1.163835e+00
std,2.402356e+01,1.693105e+01,3.490157e+00
min,-2.609000e+02,-2.968000e+01,0.000000e+00
25%,1.164000e+01,8.910000e+00,0.000000e+00
50%,1.853000e+01,1.502000e+01,0.000000e+00
75%,3.058000e+01,2.514000e+01,0.000000e+00
max,1.596100e+03,1.243910e+03,2.232500e+02


In [47]:
# Average fare and driver pay by pickup hour

hourly_financials = (
    processed_df
    .assign(pickup_hour=processed_df["pickup_datetime"].dt.hour)
    .groupby("pickup_hour")[["base_passenger_fare", "driver_pay", "tips"]]
    .mean()
    .reset_index()
)

hourly_financials

,pickup_hour,base_passenger_fare,driver_pay,tips
0,0,24.083093,19.026141,0.941940
1,1,24.735752,19.484228,0.901949
2,2,24.793534,19.936894,0.850602
3,3,25.463562,19.997011,0.883124
4,4,28.772318,21.328722,1.189486
5,5,30.053263,22.242666,1.354316
6,6,29.031999,22.046351,1.214227
7,7,27.632508,21.292321,1.101616
8,8,25.893299,20.187235,1.079669
9,9,24.699206,19.126609,1.141955


In [48]:
# Financial performance by provider

provider_financials = (
    processed_df
    .groupby("hvfhs_license_num")
    .agg(
        trip_count=("hvfhs_license_num", "size"),
        avg_fare=("base_passenger_fare", "mean"),
        avg_driver_pay=("driver_pay", "mean"),
        avg_tips=("tips", "mean")
    )
    .reset_index()
    .sort_values("trip_count", ascending=False)
)

provider_financials

,hvfhs_license_num,trip_count,avg_fare,avg_driver_pay,avg_tips
0,HV0003,15245631,26.133187,20.112076,1.123350
1,HV0005,5694742,24.360318,19.085411,1.272218


In [49]:
# Shared / pooled ride analysis

shared_analysis = (
    processed_df
    .groupby(["shared_request_flag", "shared_match_flag"])
    .agg(
        trip_count=("hvfhs_license_num", "size"),
        avg_fare=("base_passenger_fare", "mean"),
        avg_driver_pay=("driver_pay", "mean"),
        avg_tips=("tips", "mean")
    )
    .reset_index()
    .sort_values("trip_count", ascending=False)
)

shared_analysis

,shared_request_flag,shared_match_flag,trip_count,avg_fare,avg_driver_pay,avg_tips
0,N,N,20524655,25.730107,19.810232,1.179047
3,Y,Y,237619,23.545492,21.936164,0.418522
2,Y,N,176385,19.232336,19.623808,0.392826
1,N,Y,1714,31.465723,20.889936,1.677246


In [50]:
# Wheelchair-accessible (WAV) trip analysis

wav_analysis = (
    processed_df
    .groupby(["wav_request_flag", "wav_match_flag"])
    .agg(
        trip_count=("hvfhs_license_num", "size"),
        avg_fare=("base_passenger_fare", "mean"),
        avg_driver_pay=("driver_pay", "mean")
    )
    .reset_index()
    .sort_values("trip_count", ascending=False)
)

wav_analysis

,wav_request_flag,wav_match_flag,trip_count,avg_fare,avg_driver_pay
0,N,N,19019977,26.014568,19.898152
1,N,Y,1869934,22.006863,18.935329
2,Y,Y,50462,23.676972,28.488065


In [51]:
# Access-A-Ride trip analysis

access_analysis = (
    processed_df
    .groupby("access_a_ride_flag")
    .agg(
        trip_count=("hvfhs_license_num", "size"),
        avg_fare=("base_passenger_fare", "mean"),
        avg_driver_pay=("driver_pay", "mean"),
        avg_trip_miles=("trip_miles", "mean"),
        avg_trip_time=("trip_time", "mean")
    )
    .reset_index()
    .sort_values("trip_count", ascending=False)
)

access_analysis

,access_a_ride_flag,trip_count,avg_fare,avg_driver_pay,avg_trip_miles,avg_trip_time
0,N,20921620,25.645788,19.826804,4.636443,1124.890104
1,Y,18753,31.526691,26.604711,6.025099,1496.019144


In [52]:
# Airport trip analysis

airport_analysis = (
    processed_df
    .assign(airport_trip=processed_df["airport_fee"] > 0)
    .groupby("airport_trip")
    .agg(
        trip_count=("hvfhs_license_num", "size"),
        avg_fare=("base_passenger_fare", "mean"),
        avg_driver_pay=("driver_pay", "mean"),
        avg_trip_miles=("trip_miles", "mean"),
        avg_trip_time=("trip_time", "mean")
    )
    .reset_index()
)

airport_analysis

,airport_trip,trip_count,avg_fare,avg_driver_pay,avg_trip_miles,avg_trip_time
0,False,19388230,22.570838,17.868334,3.948618,1047.053891
1,True,1552143,64.126855,44.372460,13.245032,2101.646870


In [53]:
# Daily trip demand

daily_demand = (
    processed_df
    .assign(trip_date=processed_df["pickup_datetime"].dt.date)
    .groupby("trip_date")
    .size()
    .reset_index(name="trip_count")
)

daily_demand

,trip_date,trip_count
0,2026-01-01,730355
1,2026-01-02,600947
2,2026-01-03,640262
3,2026-01-04,572075
4,2026-01-05,552678
5,2026-01-06,565093
6,2026-01-07,580665
7,2026-01-08,617258
8,2026-01-09,699547
9,2026-01-10,833940


In [54]:
# Investigate the unusually low-demand day

daily_demand.loc[
    daily_demand["trip_count"].idxmin()
]

trip_date     2026-01-25
trip_count        322276
Name: 24, dtype: object

In [55]:
# Compare the low-demand day with the days around it

daily_demand[
    daily_demand["trip_date"].between(
        pd.to_datetime("2026-01-23").date(),
        pd.to_datetime("2026-01-27").date()
    )
]

,trip_date,trip_count
22,2026-01-23,804588
23,2026-01-24,880274
24,2026-01-25,322276
25,2026-01-26,439462
26,2026-01-27,682417


In [56]:
# Hourly demand on January 25

jan25_hourly = (
    processed_df[
        processed_df["pickup_datetime"].dt.date == pd.to_datetime("2026-01-25").date()
    ]
    .assign(pickup_hour=lambda x: x["pickup_datetime"].dt.hour)
    .groupby("pickup_hour")
    .size()
    .reset_index(name="trip_count")
)

jan25_hourly

,pickup_hour,trip_count
0,0,41240
1,1,31619
2,2,23742
3,3,16633
4,4,13396
5,5,10755
6,6,11337
7,7,12605
8,8,11461
9,9,10326


In [57]:
# Provider-level demand on January 25

jan25_provider = (
    processed_df[
        processed_df["pickup_datetime"].dt.date == pd.to_datetime("2026-01-25").date()
    ]
    .groupby("hvfhs_license_num")
    .size()
    .reset_index(name="trip_count")
    .sort_values("trip_count", ascending=False)
)

jan25_provider

,hvfhs_license_num,trip_count
0,HV0003,235119
1,HV0005,87157


In [58]:
# Compare provider demand around the low-demand day

provider_around_jan25 = (
    processed_df[
        processed_df["pickup_datetime"].dt.date.between(
            pd.to_datetime("2026-01-23").date(),
            pd.to_datetime("2026-01-27").date()
        )
    ]
    .assign(trip_date=lambda x: x["pickup_datetime"].dt.date)
    .groupby(["trip_date", "hvfhs_license_num"])
    .size()
    .reset_index(name="trip_count")
)

provider_around_jan25

,trip_date,hvfhs_license_num,trip_count
0,2026-01-23,HV0003,578171
1,2026-01-23,HV0005,226417
2,2026-01-24,HV0003,633500
3,2026-01-24,HV0005,246774
4,2026-01-25,HV0003,235119
5,2026-01-25,HV0005,87157
6,2026-01-26,HV0003,324502
7,2026-01-26,HV0005,114960
8,2026-01-27,HV0003,508756
9,2026-01-27,HV0005,173661


In [59]:
# Financial impact of the low-demand day

jan25_financials = (
    processed_df[
        processed_df["pickup_datetime"].dt.date == pd.to_datetime("2026-01-25").date()
    ][
        ["base_passenger_fare", "driver_pay", "tips"]
    ]
    .agg(["sum", "mean"])
)

jan25_financials

,base_passenger_fare,driver_pay,tips
sum,6.704656e+06,5.321566e+06,287859.450000
mean,2.080408e+01,1.651245e+01,0.893208


In [60]:
# Compare financial performance around January 25

financials_around_jan25 = (
    processed_df[
        processed_df["pickup_datetime"].dt.date.between(
            pd.to_datetime("2026-01-23").date(),
            pd.to_datetime("2026-01-27").date()
        )
    ]
    .assign(trip_date=lambda x: x["pickup_datetime"].dt.date)
    .groupby("trip_date")
    .agg(
        trip_count=("hvfhs_license_num", "size"),
        total_fare=("base_passenger_fare", "sum"),
        total_driver_pay=("driver_pay", "sum"),
        total_tips=("tips", "sum"),
        avg_fare=("base_passenger_fare", "mean"),
        avg_driver_pay=("driver_pay", "mean")
    )
    .reset_index()
)

financials_around_jan25

,trip_date,trip_count,total_fare,total_driver_pay,total_tips,avg_fare,avg_driver_pay
0,2026-01-23,804588,19225680.91,15390528.46,860291.41,23.895063,19.128459
1,2026-01-24,880274,19982372.97,15812580.32,870928.45,22.700174,17.963248
2,2026-01-25,322276,6704655.97,5321566.47,287859.45,20.804081,16.512450
3,2026-01-26,439462,11379801.05,8748582.65,548558.24,25.894847,19.907484
4,2026-01-27,682417,21122853.10,16366415.04,972032.28,30.953000,23.983012


In [61]:
# Investigate zero-distance trips

zero_distance_trips = processed_df[
    processed_df["trip_miles"] == 0
]

zero_distance_trips[
    [
        "pickup_datetime",
        "dropoff_datetime",
        "trip_miles",
        "trip_time",
        "base_passenger_fare",
        "driver_pay",
        "tips"
    ]
].head(20)

,pickup_datetime,dropoff_datetime,trip_miles,trip_time,base_passenger_fare,driver_pay,tips
4589,2026-01-01 00:20:49,2026-01-01 00:42:13,0.0,1284,18.24,19.34,0.0
5809,2026-01-01 00:38:49,2026-01-01 00:43:19,0.0,271,18.95,20.15,0.0
5858,2026-01-01 00:14:13,2026-01-01 00:17:36,0.0,202,8.75,7.28,0.0
8255,2026-01-01 00:48:56,2026-01-01 00:59:20,0.0,625,22.53,30.19,0.0
19690,2026-01-01 00:32:48,2026-01-01 00:36:22,0.0,214,12.70,11.08,0.0
26265,2026-01-01 00:48:31,2026-01-01 00:50:51,0.0,140,26.10,24.83,0.0
26626,2026-01-01 00:55:26,2026-01-01 01:00:12,0.0,287,10.70,8.36,0.0
27924,2026-01-01 00:14:51,2026-01-01 00:16:57,0.0,126,9.80,7.42,0.0
34914,2026-01-01 00:50:29,2026-01-01 00:51:39,0.0,71,16.88,25.90,0.0
39465,2026-01-01 00:52:32,2026-01-01 01:10:27,0.0,1075,33.03,33.15,0.0


In [62]:
# Check whether zero-distance trips have meaningful trip durations

zero_distance_summary = (
    zero_distance_trips
    .assign(
        zero_time=lambda x: x["trip_time"] == 0
    )
    .groupby("zero_time")
    .size()
    .reset_index(name="trip_count")
)

zero_distance_summary

,zero_time,trip_count
0,False,2045
1,True,737


In [63]:
# Investigate zero-distance AND zero-time trips

zero_distance_zero_time = processed_df[
    (processed_df["trip_miles"] == 0) &
    (processed_df["trip_time"] == 0)
]

zero_distance_zero_time[
    [
        "pickup_datetime",
        "dropoff_datetime",
        "PULocationID",
        "DOLocationID",
        "base_passenger_fare",
        "driver_pay",
        "tips",
        "shared_request_flag",
        "shared_match_flag"
    ]
].head(20)

,pickup_datetime,dropoff_datetime,PULocationID,DOLocationID,base_passenger_fare,driver_pay,tips,shared_request_flag,shared_match_flag
10532933,2026-01-17 07:25:39.000,2026-01-17 07:42:28.290,97,260,0.00,0.00,0.00,N,N
11356703,2026-01-18 07:32:28.000,2026-01-18 07:48:37.503,97,260,0.00,0.52,0.00,N,N
11455817,2026-01-18 11:48:19.010,2026-01-18 12:08:27.653,145,36,0.00,0.00,0.00,N,N
12018364,2026-01-19 07:44:01.000,2026-01-19 08:01:29.476,97,260,0.00,0.00,0.00,N,N
12082155,2026-01-19 10:59:49.000,2026-01-19 11:38:46.210,48,265,0.00,0.00,0.00,N,N
12505375,2026-01-19 23:31:05.000,2026-01-19 23:48:28.000,255,7,15.19,16.53,3.00,Y,N
12918167,2026-01-20 16:53:24.000,2026-01-20 17:27:15.000,62,80,27.57,27.53,0.00,N,N
12983491,2026-01-20 17:00:48.000,2026-01-20 17:27:26.000,95,112,31.37,25.61,6.98,N,N
13521695,2026-01-21 13:08:01.000,2026-01-21 13:19:19.000,92,138,26.38,14.58,0.00,N,N
13861646,2026-01-21 23:26:47.000,2026-01-21 23:37:22.530,118,84,0.00,0.00,0.00,N,N


In [64]:
zero_time_financials = zero_distance_zero_time[
    [
        "base_passenger_fare",
        "driver_pay",
        "tips"
    ]
].describe()

zero_time_financials

,base_passenger_fare,driver_pay,tips
count,737.000000,737.000000,737.000000
mean,24.767436,18.600326,0.788494
std,17.717712,13.676089,2.678562
min,0.000000,0.000000,0.000000
25%,12.520000,8.310000,0.000000
50%,19.730000,14.680000,0.000000
75%,32.300000,25.560000,0.000000
max,123.710000,104.850000,27.360000


In [65]:
zero_value_trips = zero_distance_zero_time[
    (zero_distance_zero_time["base_passenger_fare"] == 0) &
    (zero_distance_zero_time["driver_pay"] == 0) &
    (zero_distance_zero_time["tips"] == 0)
]

print("Completely zero-value trips:", len(zero_value_trips))

Completely zero-value trips: 6


### Zero-distance and zero-time trips

The dataset contains trips with `trip_miles = 0` and `trip_time = 0`.

A total of 737 records meet both conditions. Investigation shows that most of these records contain meaningful fare and driver-pay values, so they are retained rather than removed.

Only 6 records have zero distance, zero duration, zero passenger fare, zero driver pay, and zero tips. These records are treated as non-informative records and will be removed from the analytical dataset.

**Cleaning decision:** Remove 6 completely zero-value records; retain the remaining 731 zero-distance/zero-time records.

In [66]:
long_trips = processed_df[
    processed_df["trip_time"] > 14400
]

long_trips[
    [
        "pickup_datetime",
        "dropoff_datetime",
        "trip_miles",
        "trip_time",
        "base_passenger_fare",
        "driver_pay",
        "tips"
    ]
].head(20)

,pickup_datetime,dropoff_datetime,trip_miles,trip_time,base_passenger_fare,driver_pay,tips
60838,2026-01-01 01:28:32,2026-01-01 05:34:51,123.430,14779,576.05,386.29,30.00
239928,2026-01-01 04:52:04,2026-01-01 09:26:01,0.020,16437,349.53,191.11,0.00
382865,2026-01-01 11:52:02,2026-01-01 19:19:27,33.080,26845,26.19,22.45,3.19
567447,2026-01-01 18:40:42,2026-01-02 01:39:00,133.398,25098,1049.96,787.46,0.00
670864,2026-01-01 21:20:09,2026-01-02 03:58:18,340.480,23889,738.28,685.80,0.00
717116,2026-01-01 23:23:18,2026-01-02 05:09:14,225.130,20756,634.86,623.58,0.00
775443,2026-01-02 04:12:39,2026-01-02 08:50:15,17.230,16656,308.47,208.07,9.89
913510,2026-01-02 11:23:16,2026-01-02 16:27:04,17.340,18229,203.50,222.48,0.00
941455,2026-01-02 12:17:35,2026-01-02 20:22:54,145.698,29119,1179.40,884.54,0.00
949160,2026-01-02 12:37:44,2026-01-02 16:53:21,95.010,15337,615.06,390.49,0.00


In [67]:
long_trip_summary = long_trips.assign(
    avg_speed_mph=long_trips["trip_miles"] / (long_trips["trip_time"] / 3600)
)[
    [
        "trip_miles",
        "trip_time",
        "avg_speed_mph"
    ]
].describe()

long_trip_summary

,trip_miles,trip_time,avg_speed_mph
count,255.00000,255.000000,255.000000
mean,109.34411,18893.854902,21.529379
std,108.51305,5829.995946,20.800915
min,0.00000,14408.000000,0.000000
25%,8.03500,15043.000000,1.335320
50%,72.83800,16274.000000,12.504430
75%,205.28500,19835.500000,44.035967
max,571.31000,44998.000000,63.580899


In [68]:
id="low_speed_long_trips"
low_speed_long_trips = (
    long_trips.assign(
        avg_speed_mph=long_trips["trip_miles"] / (long_trips["trip_time"] / 3600)
    )
    .sort_values("avg_speed_mph")
)

low_speed_long_trips[
    [
        "pickup_datetime",
        "dropoff_datetime",
        "trip_miles",
        "trip_time",
        "avg_speed_mph",
        "base_passenger_fare",
        "driver_pay",
        "tips"
    ]
].head(20)

,pickup_datetime,dropoff_datetime,trip_miles,trip_time,avg_speed_mph,base_passenger_fare,driver_pay,tips
8575416,2026-01-14 14:17:00,2026-01-14 20:09:55,0.000,21176,0.000000,232.44,153.14,0.0
7457797,2026-01-12 17:31:46,2026-01-12 21:33:25,0.000,14499,0.000000,-16.69,3.48,0.0
1250447,2026-01-02 21:19:04,2026-01-03 01:26:55,0.000,14871,0.000000,-16.15,24.92,0.0
239928,2026-01-01 04:52:04,2026-01-01 09:26:01,0.020,16437,0.004380,349.53,191.11,0.0
16869054,2026-01-26 18:00:45,2026-01-26 22:02:45,0.030,14519,0.007439,0.50,19.36,0.0
8561216,2026-01-14 13:54:27,2026-01-14 17:58:56,0.080,14669,0.019633,95.00,147.86,0.0
8667960,2026-01-14 17:15:05,2026-01-14 21:19:03,0.090,14637,0.022136,-11.36,12.36,0.0
3513857,2026-01-06 18:07:24,2026-01-06 22:15:05,0.130,14862,0.031490,170.73,163.85,0.0
6493301,2026-01-11 03:41:34,2026-01-11 08:57:42,0.210,18968,0.039857,284.64,208.60,0.0
8631195,2026-01-14 16:55:08,2026-01-14 20:56:53,0.230,14504,0.057088,-4.58,40.83,0.0


In [69]:
extreme_slow_trips = long_trips.assign(
    avg_speed_mph=long_trips["trip_miles"] / (long_trips["trip_time"] / 3600)
)

print("Long trips:", len(extreme_slow_trips))
print("Speed < 1 mph:", (extreme_slow_trips["avg_speed_mph"] < 1).sum())
print("Speed < 0.5 mph:", (extreme_slow_trips["avg_speed_mph"] < 0.5).sum())
print("Speed = 0 mph:", (extreme_slow_trips["avg_speed_mph"] == 0).sum())

Long trips: 255
Speed < 1 mph: 56
Speed < 0.5 mph: 30
Speed = 0 mph: 3


In [70]:
extreme_slow_trips[
    extreme_slow_trips["avg_speed_mph"] < 1
][[
    "pickup_datetime",
    "dropoff_datetime",
    "trip_miles",
    "trip_time",
    "avg_speed_mph",
    "base_passenger_fare",
    "driver_pay",
    "tips",
    "hvfhs_license_num"
]].sort_values("avg_speed_mph").head(20)

,pickup_datetime,dropoff_datetime,trip_miles,trip_time,avg_speed_mph,base_passenger_fare,driver_pay,tips,hvfhs_license_num
1250447,2026-01-02 21:19:04,2026-01-03 01:26:55,0.000,14871,0.000000,-16.15,24.92,0.0,HV0003
8575416,2026-01-14 14:17:00,2026-01-14 20:09:55,0.000,21176,0.000000,232.44,153.14,0.0,HV0003
7457797,2026-01-12 17:31:46,2026-01-12 21:33:25,0.000,14499,0.000000,-16.69,3.48,0.0,HV0003
239928,2026-01-01 04:52:04,2026-01-01 09:26:01,0.020,16437,0.004380,349.53,191.11,0.0,HV0003
16869054,2026-01-26 18:00:45,2026-01-26 22:02:45,0.030,14519,0.007439,0.50,19.36,0.0,HV0003
8561216,2026-01-14 13:54:27,2026-01-14 17:58:56,0.080,14669,0.019633,95.00,147.86,0.0,HV0003
8667960,2026-01-14 17:15:05,2026-01-14 21:19:03,0.090,14637,0.022136,-11.36,12.36,0.0,HV0003
3513857,2026-01-06 18:07:24,2026-01-06 22:15:05,0.130,14862,0.031490,170.73,163.85,0.0,HV0003
6493301,2026-01-11 03:41:34,2026-01-11 08:57:42,0.210,18968,0.039857,284.64,208.60,0.0,HV0003
8631195,2026-01-14 16:55:08,2026-01-14 20:56:53,0.230,14504,0.057088,-4.58,40.83,0.0,HV0003


In [71]:
originating_missing = (
    processed_df[processed_df["originating_base_num"].isna()]
    .groupby("hvfhs_license_num")
    .size()
    .reset_index(name="missing_count")
    .sort_values("missing_count", ascending=False)
)

originating_missing

,hvfhs_license_num,missing_count
0,HV0005,5676843


In [72]:
# Check for logically invalid trip timestamps

invalid_timestamps = processed_df[
    (processed_df["dropoff_datetime"] < processed_df["pickup_datetime"]) |
    (processed_df["pickup_datetime"] < processed_df["request_datetime"])
]

print("Trips with invalid timestamp order:", len(invalid_timestamps))

Trips with invalid timestamp order: 267757


In [73]:
timestamp_issues = pd.Series({
    "dropoff_before_pickup": (
        processed_df["dropoff_datetime"] < processed_df["pickup_datetime"]
    ).sum(),

    "pickup_before_request": (
        processed_df["pickup_datetime"] < processed_df["request_datetime"]
    ).sum(),

    "both_issues": (
        (processed_df["dropoff_datetime"] < processed_df["pickup_datetime"]) &
        (processed_df["pickup_datetime"] < processed_df["request_datetime"])
    ).sum()
})

timestamp_issues

dropoff_before_pickup         0
pickup_before_request    267757
both_issues                   0
dtype: int64

In [74]:
physical_checks = pd.Series({
    "negative_trip_miles": (processed_df["trip_miles"] < 0).sum(),
    "negative_trip_time": (processed_df["trip_time"] < 0).sum(),
    "negative_tolls": (processed_df["tolls"] < 0).sum(),
    "negative_sales_tax": (processed_df["sales_tax"] < 0).sum(),
    "negative_congestion_surcharge": (processed_df["congestion_surcharge"] < 0).sum(),
    "negative_airport_fee": (processed_df["airport_fee"] < 0).sum()
})

physical_checks

negative_trip_miles              0
negative_trip_time               0
negative_tolls                   0
negative_sales_tax               0
negative_congestion_surcharge    0
negative_airport_fee             0
dtype: int64

In [75]:
# Create the cleaned analytical dataset
cleaned_df = processed_df.drop(index=zero_value_trips.index).copy()

print("Original rows:", len(processed_df))
print("Removed rows:", len(processed_df) - len(cleaned_df))
print("Cleaned rows:", len(cleaned_df))

Original rows: 20940373
Removed rows: 6
Cleaned rows: 20940367


In [76]:
# Save cleaned January dataset

processed_data = project / "processed data"
processed_data.mkdir(exist_ok=True)

cleaned_file = processed_data / "fhvhv_tripdata_2026-01_cleaned.parquet"

cleaned_df.to_parquet(cleaned_file, index=False)

print("Saved to:", cleaned_file)
print("File exists:", cleaned_file.exists())

Saved to: /Users/dikshadev/    Projects/NYC_HV_FHV_ Analytics_2026/processed data/fhvhv_tripdata_2026-01_cleaned.parquet
File exists: True


# Business EDA

## 1. Demand by Hour

Analyze how trip demand varies throughout the day to identify peak and off-peak hours.

In [77]:
# Hourly trip demand

hourly_demand = (
    cleaned_df
    .assign(pickup_hour=cleaned_df["pickup_datetime"].dt.hour)
    .groupby("pickup_hour")
    .size()
    .reset_index(name="trip_count")
    .sort_values("pickup_hour")
)

hourly_demand

,pickup_hour,trip_count
0,0,729999
1,1,515768
2,2,386609
3,3,313318
4,4,321638
5,5,377950
6,6,582924
7,7,886281
8,8,1087379
9,9,990742


## 2. Demand by Day

Analyze daily trip volume throughout January 2026 to identify unusually high- and low-demand days.

In [78]:
# Daily trip demand

daily_demand = (
    cleaned_df
    .assign(trip_date=cleaned_df["pickup_datetime"].dt.date)
    .groupby("trip_date")
    .size()
    .reset_index(name="trip_count")
    .sort_values("trip_date")
)

daily_demand

,trip_date,trip_count
0,2026-01-01,730355
1,2026-01-02,600947
2,2026-01-03,640262
3,2026-01-04,572075
4,2026-01-05,552678
5,2026-01-06,565093
6,2026-01-07,580665
7,2026-01-08,617258
8,2026-01-09,699547
9,2026-01-10,833940


## 3. Demand by Day of Week

Analyze trip demand by day of week to identify recurring weekday and weekend demand patterns.

In [79]:
# Demand by day of week

weekday_demand = (
    cleaned_df
    .assign(
        day_of_week=cleaned_df["pickup_datetime"].dt.day_name()
    )
    .groupby("day_of_week")
    .size()
    .reindex([
        "Monday",
        "Tuesday",
        "Wednesday",
        "Thursday",
        "Friday",
        "Saturday",
        "Sunday"
    ])
    .reset_index(name="trip_count")
)

weekday_demand

,day_of_week,trip_count
0,Monday,2199131
1,Tuesday,2536803
2,Wednesday,2606896
3,Thursday,3508966
4,Friday,3754035
5,Saturday,4045716
6,Sunday,2288820


## 4. Provider Market Share

Analyze the distribution of January trips across HVFHS providers to understand provider-level market share.

In [80]:
# Provider-level trip volume and market share

provider_demand = (
    cleaned_df
    .groupby("hvfhs_license_num")
    .size()
    .reset_index(name="trip_count")
)

provider_demand["market_share_pct"] = (
    provider_demand["trip_count"]
    / provider_demand["trip_count"].sum()
    * 100
)

provider_demand.sort_values(
    "trip_count",
    ascending=False
)

,hvfhs_license_num,trip_count,market_share_pct
0,HV0003,15245625,72.804956
1,HV0005,5694742,27.195044


## 5. Provider Financial Performance

Compare average fare, driver pay, and tips across providers to understand differences in trip-level financial performance.

In [81]:
# Provider-level financial performance

provider_financials = (
    cleaned_df
    .groupby("hvfhs_license_num")
    .agg(
        trip_count=("hvfhs_license_num", "size"),
        avg_fare=("base_passenger_fare", "mean"),
        avg_driver_pay=("driver_pay", "mean"),
        avg_tips=("tips", "mean")
    )
    .reset_index()
    .sort_values("trip_count", ascending=False)
)

provider_financials

,hvfhs_license_num,trip_count,avg_fare,avg_driver_pay,avg_tips
0,HV0003,15245625,26.133197,20.112084,1.123351
1,HV0005,5694742,24.360318,19.085411,1.272218


## 6. Trip Distance Distribution

Analyze trips by distance bands to understand the typical travel range and the concentration of short- and long-distance trips.

In [82]:
# Trip distance distribution

distance_bands = pd.cut(
    cleaned_df["trip_miles"],
    bins=[-0.01, 1, 3, 5, 10, 20, float("inf")],
    labels=[
        "0–1 miles",
        "1–3 miles",
        "3–5 miles",
        "5–10 miles",
        "10–20 miles",
        "20+ miles"
    ]
)

distance_distribution = (
    distance_bands
    .value_counts()
    .sort_index()
    .reset_index()
)

distance_distribution.columns = ["distance_band", "trip_count"]

distance_distribution["share_pct"] = (
    distance_distribution["trip_count"]
    / distance_distribution["trip_count"].sum()
    * 100
)

distance_distribution

,distance_band,trip_count,share_pct
0,0–1 miles,2768895,13.222763
1,1–3 miles,8530788,40.738484
2,3–5 miles,3579086,17.091802
3,5–10 miles,3662790,17.491527
4,10–20 miles,1985377,9.481099
5,20+ miles,413431,1.974325


## 7. Trip Duration Distribution

Analyze trips by duration bands to understand typical trip lengths and the concentration of short- and long-duration rides.

In [83]:
# Trip duration distribution

duration_bands = pd.cut(
    cleaned_df["trip_time"] / 60,
    bins=[0, 5, 10, 20, 30, 60, float("inf")],
    labels=[
        "0–5 min",
        "5–10 min",
        "10–20 min",
        "20–30 min",
        "30–60 min",
        "60+ min"
    ],
    include_lowest=True
)

duration_distribution = (
    duration_bands
    .value_counts()
    .sort_index()
    .reset_index()
)

duration_distribution.columns = ["duration_band", "trip_count"]

duration_distribution["share_pct"] = (
    duration_distribution["trip_count"]
    / duration_distribution["trip_count"].sum()
    * 100
)

duration_distribution

,duration_band,trip_count,share_pct
0,0–5 min,1127720,5.385388
1,5–10 min,4609129,22.010736
2,10–20 min,7876748,37.615138
3,20–30 min,4051476,19.347684
4,30–60 min,2970137,14.183787
5,60+ min,305157,1.457267


## 8. Pickup Location Demand

Analyze trip volume by pickup location to identify areas with the highest concentration of HVFHV demand.

In [84]:
# Pickup location demand

pickup_location_demand = (
    cleaned_df
    .groupby("PULocationID")
    .size()
    .reset_index(name="trip_count")
    .sort_values("trip_count", ascending=False)
)

pickup_location_demand["share_pct"] = (
    pickup_location_demand["trip_count"]
    / pickup_location_demand["trip_count"].sum()
    * 100
)

pickup_location_demand.head(15)

,PULocationID,trip_count,share_pct
135,138,364872,1.742434
129,132,330578,1.578664
60,61,277903,1.327116
36,37,246773,1.178456
75,76,245850,1.174048
78,79,242997,1.160424
227,230,231846,1.107173
158,161,226497,1.081629
178,181,214217,1.022986
228,231,213850,1.021233


## 9. Most Common Pickup-to-Drop-off Routes

Analyze origin-destination combinations to identify the most frequently used travel corridors in the HVFHV network.

In [85]:
# Most common pickup-to-drop-off routes

route_demand = (
    cleaned_df
    .groupby(["PULocationID", "DOLocationID"])
    .size()
    .reset_index(name="trip_count")
    .sort_values("trip_count", ascending=False)
)

route_demand["share_pct"] = (
    route_demand["trip_count"]
    / route_demand["trip_count"].sum()
    * 100
)

route_demand.head(20)

,PULocationID,DOLocationID,trip_count,share_pct
16645,76,76,83414,0.398341
28705,132,265,63675,0.304078
5159,26,26,57540,0.274780
8016,39,39,50486,0.241094
30165,138,265,45180,0.215756
2278,14,14,40512,0.193464
12957,61,61,37757,0.180307
21243,95,95,35166,0.167934
833,7,7,34732,0.165861
27825,129,129,34595,0.165207


## 10. Hourly Revenue and Driver Pay

Analyze average passenger fare and driver pay by pickup hour to understand how trip-level financial outcomes vary throughout the day.

In [86]:
# Hourly financial performance

hourly_financials = (
    cleaned_df
    .assign(pickup_hour=cleaned_df["pickup_datetime"].dt.hour)
    .groupby("pickup_hour")
    .agg(
        trip_count=("pickup_hour", "size"),
        avg_fare=("base_passenger_fare", "mean"),
        avg_driver_pay=("driver_pay", "mean"),
        avg_tips=("tips", "mean")
    )
    .reset_index()
)

hourly_financials

,pickup_hour,trip_count,avg_fare,avg_driver_pay,avg_tips
0,0,729999,24.083093,19.026141,0.941940
1,1,515768,24.735752,19.484228,0.901949
2,2,386609,24.793534,19.936894,0.850602
3,3,313318,25.463562,19.997011,0.883124
4,4,321638,28.772318,21.328722,1.189486
5,5,377950,30.053263,22.242666,1.354316
6,6,582924,29.031999,22.046351,1.214227
7,7,886281,27.632602,21.292393,1.101620
8,8,1087379,25.893299,20.187235,1.079669
9,9,990742,24.699206,19.126609,1.141955


## 11. Shared Ride Adoption

Analyze shared-ride requests and successful shared matches to understand the adoption of pooled trips within the HVFHV network.

In [89]:
# Shared ride adoption

total_trips = len(cleaned_df)

shared_requests = (cleaned_df["shared_request_flag"] == "Y").sum()

shared_matches = (cleaned_df["shared_match_flag"] == "Y").sum()

shared_ride = pd.DataFrame({
    "total_trips": [total_trips],
    "shared_requests": [shared_requests],
    "shared_matches": [shared_matches]
})

shared_ride["request_rate_pct"] = (
    shared_ride["shared_requests"]
    / shared_ride["total_trips"]
    * 100
)

shared_ride["match_rate_pct"] = (
    shared_ride["shared_matches"]
    / shared_ride["shared_requests"]
    * 100
)

shared_ride

,total_trips,shared_requests,shared_matches,request_rate_pct,match_rate_pct
0,20940367,414004,239333,1.977062,57.809345


## 12. Airport-Related Trip Analysis

Analyze the prevalence of airport-related trips using the airport fee field to understand the importance of airport travel within the HVFHV network.

In [90]:
# Airport-related trips

airport_trips = pd.DataFrame({
    "total_trips": [len(cleaned_df)],
    "airport_trips": [(cleaned_df["airport_fee"] > 0).sum()]
})

airport_trips["airport_trip_rate_pct"] = (
    airport_trips["airport_trips"]
    / airport_trips["total_trips"]
    * 100
)

airport_trips

,total_trips,airport_trips,airport_trip_rate_pct
0,20940367,1552143,7.412205
